sound/soc/sof/intel/mtl.c

```diff
@@ -612,6 +613,18 @@ static int mtl_dsp_disable_interrupts(struct snd_sof_dev *sdev)
 	return mtl_disable_interrupts(sdev);
 }
 
+static u64 mtl_dsp_get_stream_hda_link_position(struct snd_sof_dev *sdev,
+						struct snd_soc_component *component,
+						struct snd_pcm_substream *substream)
+{
+	struct hdac_stream *hstream = substream->runtime->private_data;
+	u32 llp_l, llp_u;
+
+	llp_l = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, MTL_PPLCLLPL(hstream->index));
+	llp_u = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, MTL_PPLCLLPU(hstream->index));
+	return ((u64)llp_l | (u64)llp_u << 32);
```

Comment:
I am still in the dark on how we would do timestamping if there is a mixer in the firmware. You will use the information for the link DMA which may have been started already. Have you tested this with the deep-buffer addition for HDaudio?

Should we report a zero-delay in case we have more than one FE connected to the BE?

Reply:
yes, I tested on mtl with closed source fw. For mixer case, two hda streams have the same llp counter read in host space but have different stream_start_offset in fw.  stream_start_offset is substracted from llp value for calculation. As I mentioned, stream_start_offset is calculated by fw independent of hw counter in fw, some sw counters like gateway processed sample bytes and  pipeline latency.